# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/layers/SurfacePoolWayland.cpp
# Commit: 086a82512429
# Full Hash: 086a825124296adeaf48cc3d05975e82f55fcb85
# Author: Robert Mader <robert.mader@posteo.de>
# Date: 2021-06-21 09:42:23
# Description:
#   Bug 1716006 - Clear pending callbacks when destroying a NativeSurfaceWayland, r=gfx-reviewers,gw
#   
#   Unlike buffer release callbacks, which can't happen after the
#   corresponding buffer was destroyed, frame callbacks can apparently
#   arrive even if the corresponding surface was destroyed.
# ==============================================================================

diff -r 95970359b68e -r 086a82512429 gfx/layers/SurfacePoolWayland.cpp
--- a/gfx/layers/SurfacePoolWayland.cpp	Sun Jun 20 15:46:38 2021 +0000
+++ b/gfx/layers/SurfacePoolWayland.cpp	Sun Jun 20 21:05:47 2021 +0000
@@ -86,6 +86,8 @@
 }
 
 NativeSurfaceWayland::~NativeSurfaceWayland() {
+  MutexAutoLock lock(mMutex);
+  g_clear_pointer(&mCallback, wl_callback_destroy);
   g_clear_pointer(&mViewport, wp_viewport_destroy);
   g_clear_pointer(&mWlSubsurface, wl_subsurface_destroy);
   g_clear_pointer(&mWlSurface, wl_surface_destroy);
@@ -167,12 +169,11 @@
       [&](const auto& object) { return !object->IsActive(); });
 
   mCallbackMultiplexHelpers.AppendElement(aMultiplexHelper);
-  if (!mCallbackRequested) {
-    wl_callback* callback = wl_surface_frame(mWlSurface);
-    wl_callback_add_listener(callback, &sFrameListenerNativeSurfaceWayland,
+  if (!mCallback) {
+    mCallback = wl_surface_frame(mWlSurface);
+    wl_callback_add_listener(mCallback, &sFrameListenerNativeSurfaceWayland,
                              this);
     wl_surface_commit(mWlSurface);
-    mCallbackRequested = true;
   }
 }
 
@@ -180,8 +181,8 @@
                                                 uint32_t aTime) {
   MutexAutoLock lock(mMutex);
 
-  wl_callback_destroy(aCallback);
-  mCallbackRequested = false;
+  MOZ_RELEASE_ASSERT(aCallback == mCallback);
+  g_clear_pointer(&mCallback, wl_callback_destroy);
 
   for (const RefPtr<CallbackMultiplexHelper>& callbackMultiplexHelper :
        mCallbackMultiplexHelpers) {